cpp/examples/arrow/exec_plan_examples.cc

```diff
@@ -0,0 +1,1122 @@
+
+    // pipe the scan node into a filter node
+    cp::ExecNode* filter;
+    ARROW_ASSIGN_OR_RAISE(filter, cp::MakeExecNode("filter", plan.get(), {scan},
+                                                    cp::FilterNodeOptions{b_is_true}));
+
+    cp::ExecNode* project;
+
+    ARROW_ASSIGN_OR_RAISE(project,
+                            cp::MakeExecNode("augmented_project", plan.get(), {filter},
+                                            cp::ProjectNodeOptions{{a_times_2}}));
+
+    // // finally, pipe the project node into a sink node
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+    ARROW_ASSIGN_OR_RAISE(cp::ExecNode * sink,
+                            cp::MakeExecNode("sink",
+                            plan.get(), {project},
+                            cp::SinkNodeOptions{&sink_gen}));
+
+    ABORT_ON_FAILURE(sink->Validate());
+
+    // // translate sink_gen (async) to sink_reader (sync)
+    std::shared_ptr<arrow::RecordBatchReader> sink_reader =
+        cp::MakeGeneratorReader(arrow::schema({arrow::field("a * 2", arrow::int32())}),
+                                std::move(sink_gen), exec_context.memory_pool());
+
+    // // validate the plan
+    ABORT_ON_FAILURE(plan->Validate());
+    PRINT_LINE("Exec Plan created: " << plan->ToString());
+    // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
+
+    // // collect sink_reader into a Table
+    std::shared_ptr<arrow::Table> response_table;
+    ARROW_ASSIGN_OR_RAISE(response_table,
+                            arrow::Table::FromRecordBatchReader(sink_reader.get()));
+
+    std::cout << "Results : " << response_table->ToString() << std::endl;
+
+    // // stop producing
+    plan->StopProducing();
+
+    // // plan mark finished
+    plan->finished().Wait();
+
+    return arrow::Status::OK();
+}
+
+cp::Expression Materialize(std::vector<std::string> names,
+                           bool include_aug_fields = false) {
+    if (include_aug_fields) {
+        for (auto aug_name : {"__fragment_index",
+            "__batch_index", "__last_in_fragment"}) {
+        names.emplace_back(aug_name);
+        }
+    }
+
+    std::vector<cp::Expression> exprs;
+    for (const auto& name : names) {
+        exprs.push_back(cp::field_ref(name));
+    }
+
+    return cp::project(exprs, names);
+}
+
+arrow::Status consume(std::shared_ptr<arrow::Schema> schema,
+    std::function<arrow::Future<arrow::util::optional<cp::ExecBatch>>()>* sink_gen) {
+    auto iterator = MakeGeneratorIterator(*sink_gen);
+    while (true) {
+        ARROW_ASSIGN_OR_RAISE(auto exec_batch, iterator.Next());
+        if (!exec_batch.has_value()) {
+            break;
+        }
+        ARROW_ASSIGN_OR_RAISE(auto record_batch, exec_batch->ToRecordBatch(schema));
+        std::cout << record_batch->ToString() << '\n';
+    }
+    return arrow::Status::OK();
+}
+
+
+arrow::Status scan_sink_node_example() {
```

Comment:
nit, but usually we use UpperCamelCase for function names ("Consume", "ScanSinkNodeExample"). We use snake_case only for const getters.

Reply:
Updated Naming convention